include/envoy/router/router.h

```diff
@@ -969,6 +995,23 @@ class Config {
                                     const StreamInfo::StreamInfo& stream_info,
                                     uint64_t random_value) const PURE;
 
+  /**
+   * Based on the incoming HTTP request headers, determine the target route (containing either a
+   * route entry or a direct response entry) for the request.
+   *
+   * Invokes callback with matched route, callback can choose to accept the route by returning
+   * RouteStatus::Stop or continue route match from last matched route by returning
+   * RouteMatchStatus::Continue, when more routes available.
+   *
+   * @param cb supplies callback to be invoked upon route match.
+   * @param headers supplies the request headers.
+   * @param random_value supplies the random seed to use if a runtime choice is required. This
+   *        allows stable choices between calls if desired.
+   */
+
+  virtual void route(const RouteCallback& cb, const Http::RequestHeaderMap& headers,
```

Comment:
For consistency I would have this return the route. Then the implementation is simpler to merge.